File: libraries/ndf/ndfFtest.c
#include "sae_par.h"
#include "dat_par.h"
#include "ndf1.h"
#include "ndf_err.h"
#include "ndf.h"
#include "mers.h"

int main(){
/*
*+
*  Name:
*     ndfTest

*  Purpose:
*     Test installation of NDF From Fortran.

*  Description:
*     This program should be run after building and installing NDF in order
*     to test for correct installation. Note that this is not an exhaustive
*     test of NDF, but only of its installation.

*  Licence:
*     This program is free software; you can redistribute it and/or modify
*     it under the terms of the GNU General Public License as published by
*     the Free Software Foundation; either version 2 of the License, or (at
*     your option) any later version.
*
*     This program is distributed in the hope that it will be useful,but
*     WITHOUT ANY WARRANTY; without even the implied warranty of
*     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
*     General Public License for more details.
*
*     You should have received a copy of the GNU General Public License
*     along with this program; if not, write to the Free Software
*     Foundation, Inc., 51 Franklin Street,Fifth Floor, Boston, MA
*     02110-1301, USA

*  Authors:
*     RFWS: R.F. Warren-Smith (STARLINK)
*     DSB: David S. Berry (EAO)

*  Copyright:
*     Copyright (C) 2018 East Asian Observatory
*     All rights reserved.

*  History:
*     xxx (DSB):
*        Original version, based on equivalent Fortran function by RFWS.

*-
*/

/* Local Variables: */
   int dim[ 2 ] = { 10, 20 };      /* NDF dimensions */
   int el;               /* Number of mapped elements */
   int indf;             /* NDF identifier */
   int isum;             /* Sum of array elements */
   int place;            /* NDF placeholder */
   void *pntr;           /* Pointer to mapped array */

/* Initialise the global status. */
   status = SAI__OK;

/* Create a new file containing an NDF. */
   ndfOpen( NULL, "ndf_test", "WRITE", "NEW", indf, place, status );
   ndfNewp( "_INTEGER", 2, dim, place, indf, status );

/* Map the NDF's data array. */
   ndfMap( indf, "Data", "_REAL", "WRITE", pntr, el, status );

/* Initialise the array. */
   setup( el, pntr, status );

/* Clean up. */
   ndfAnnul( indf, status );

/* Re-open the NDF. */
   ndfOpen( NULL, "ndf_test", "UPDATE", "OLD", indf, place, status );

/* Map its data array. */
   ndfMap( indf, "Data", "_INTEGER", "READ", pntr, el, status );

/* Sum the data elements. */
   sum( el, pntr, isum, status );

/* Clean up, deleting the NDF. */
   ndfDelet( indf, status );

/* Check if the test ran OK. If so, then report success. */
   if( ( status == SAI__OK ) && ( isum == 20100 ) ) {
      printf( "*********************************************\n");
      printf( "*                                           *\n");
      printf( "*  NDF Fortran installation test succeeded  *\n");
      printf( "*                                           *\n");
      printf( "*********************************************\n");

/* Otherwise, report an error. */
   } else {
      if( status == SAI__OK ) status = SAI__ERROR;
      errRep( " ", "ndfTest_F: NDF Fortran installation test failed.", status );
   }

}

void setup( int el, float array[], int *status ){
/*
*+
*  Name:
*     setup

*  Purpose:
*     Initialise an array.

*  Synopsis:
*     void setup( int el, float array[], int *status )

*  Description:
*     Set each element of a 1-dimensional array equal to its element
*     number.

*  Parameters:
*     el
*        Number of array elements.
*     array
*        Returned holding the array to be initialised.
*     *status
*        The global status.

*  Authors:
*     RFWS: R.F. Warren-Smith (STARLINK)
*     DSB: David S. Berry (EAO)

*  History:
*     xxx (DSB):
*        Original version, based on equivalent Fortran function by RFWS.

*-
*/

/* Local Variables: */
   int i;                /* Loop counter */

/* Check inherited global status. */
   if( *status != SAI__OK ) return;

/* Initialise the array. */
   for( i = 0; i < el; i++ ){
      array[ i ] = real( i + 1 );
   }

}


void sum( int el, const int array[], int *isum, int *status ){
/*
*+
*  Name:
*     sum

*  Purpose:
*     Sum the elements of an array.

*  Synopsis:
*     void sum( int el, const int array[], int *isum, int *status )

*  Description:
*     Return the sum of the elements of a 1-dimensional array.

*  Parameters:
*     el
*        Number of array elements.
*     array
*        Array whose elements are to be summed.
*     *isum
*        Returned holding the sum of array elements.
*     *status
*        The global status.

*  Authors:
*     RFWS: R.F. Warren-Smith (STARLINK)
*     DSB: David S. Berry (EAO)

*  History:
*     xxx (DSB):
*        Original version, based on equivalent Fortran function by RFWS.

*-
*/

/* Local Variables: */
   int i;                /* Loop counter */

/* Check inherited global status. */
   if( *status != SAI__OK ) return;

/* Initialise. */
   *isum = 0;

/* Sum the array elements. */
   for( i = 0; i < el; i++ ){
      *isum += array[ i ];
   }

}

File: libraries/ndf/ndf_ftest.f
      PROGRAM NDF_TEST
*+
*  Name:
*     NDF_TEST (Fortran version)

*  Purpose:
*     Test installation of NDF From Fortran.

*  Language:
*     Starlink Fortran 77

*  Description:
*     This program should be run after building and installing NDF in
*     order to test for correct installation. Note that this is not an
*     exhaustive test of NDF, but only of its installation.

*  Licence:
*     This program is free software; you can redistribute it and/or
*     modify it under the terms of the GNU General Public License as
*     published by the Free Software Foundation; either version 2 of
*     the License, or (at your option) any later version.
*
*     This program is distributed in the hope that it will be
*     useful,but WITHOUT ANY WARRANTY; without even the implied
*     warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
*     PURPOSE. See the GNU General Public License for more details.
*
*     You should have received a copy of the GNU General Public License
*     along with this program; if not, write to the Free Software
*     Foundation, Inc., 51 Franklin Street,Fifth Floor, Boston, MA
*     02110-1301, USA

*  Authors:
*     RFWS: R.F. Warren-Smith (STARLINK, RAL)
*     {enter_new_authors_here}

*  Copyright:
*     Copyright (C) 1998 Central Laboratory of the Research Councils

*  History:
*     16-JAN-1992 (RFWS):
*        Original version, derived from the equivalent HDS routine.
*     7-SEP-1993 (RFWS):
*        Remove calls to HDS_START and HDS_STOP (no longer needed) and
*        use DAT_ANNUL instead of HDS_CLOSE.
*     8-SEP-1993 (RFWS):
*        Changed to use NDF_OPEN.
*     {enter_further_changes_here}

*  Bugs:
*     {note_any_bugs_here}

*-

*  Type Definitions:
      IMPLICIT NONE              ! No implicit typing

*  Global Constants:
      INCLUDE 'SAE_PAR'          ! Standard SAE constants
      INCLUDE 'CNF_PAR'          ! For CNF_PVAL function
      INCLUDE 'DAT_PAR'          ! DAT_ public constants
      INCLUDE 'NDF_PAR'          ! NDF_ public constants
      INCLUDE 'NDF_ERR'          ! NDF_ error codes

*  Status:
      INTEGER STATUS             ! Global status

*  Local Variables:
      INTEGER DIM( 2 )           ! NDF dimensions
      INTEGER EL                 ! Number of mapped elements
      INTEGER INDF               ! NDF identifier
      INTEGER ISUM               ! Sum of array elements
      INTEGER PLACE              ! NDF placeholder
      INTEGER PNTR               ! Pointer to mapped array

*  External References:
      INCLUDE 'NDF_FUNC'         ! NDF statement functions
                                 ! (test existence)

*  Local Data:
      DATA DIM / 10, 20 /

*.

*  Initialise the global status.
      STATUS = SAI__OK

*  Create a new file containing an NDF.
      CALL NDF_OPEN( DAT__ROOT, 'ndf_test', 'WRITE', 'NEW', INDF, PLACE,
     :               STATUS )
      CALL NDF_NEWP( '_INTEGER', 2, DIM, PLACE, INDF, STATUS )

*  Map the NDF's data array.
      CALL NDF_MAP( INDF, 'Data', '_REAL', 'WRITE', PNTR, EL, STATUS )

*  Initialise the array.
      CALL SETUP( EL, %VAL( CNF_PVAL( PNTR ) ), STATUS )

*  Clean up.
      CALL NDF_ANNUL( INDF, STATUS )

*  Re-open the NDF.
      CALL NDF_OPEN( DAT__ROOT, 'ndf_test', 'UPDATE', 'OLD', INDF,
     :               PLACE, STATUS )

*  Map its data array.
      CALL NDF_MAP( INDF, 'Data', '_INTEGER', 'READ', PNTR, EL, STATUS )

*  Sum the data elements.
      CALL SUM( EL, %VAL( CNF_PVAL( PNTR ) ), ISUM, STATUS )

*  Clean up, deleting the NDF.
      CALL NDF_DELET( INDF, STATUS )

*  Check if the test ran OK. If so, then report success.
      IF ( ( STATUS .EQ. SAI__OK ) .AND. ( ISUM .EQ. 20100 ) ) THEN
         WRITE( *, * ) '*********************************************'
         WRITE( *, * ) '*                                           *'
         WRITE( *, * ) '*  NDF Fortran installation test succeeded  *'
         WRITE( *, * ) '*                                           *'
         WRITE( *, * ) '*********************************************'

*  Otherwise, report an error.
      ELSE
         IF ( STATUS .EQ. SAI__OK ) STATUS = SAI__ERROR
         CALL ERR_REP( 'NDF_TEST_ERR',
     :   'NDF_TEST_F: NDF Fortran installation test failed.', STATUS )
      END IF

      END

      SUBROUTINE SETUP( EL, ARRAY, STATUS )
*+
*  Name:
*     SETUP

*  Purpose:
*     Initialise an array.

*  Language:
*     Starlink Fortran 77

*  Invocation:
*     CALL SETUP( EL, ARRAY, STATUS )

*  Description:
*     Set each element of a 1-dimensional array equal to its element
*     number.

*  Arguments:
*     EL = INTEGER (Given)
*        Number of array elements.
*     ARRAY( EL ) = REAL (Returned)
*        Array to be initialised.
*     STATUS = INTEGER (Given and Returned)
*        The global status.

*  Authors:
*     RFWS: R.F. Warren-Smith (STARLINK, RAL)
*     {enter_new_authors_here}

*  History:
*     19-AUG-1991 (RFWS):
*        Original version.
*     {enter_changes_here}

*  Bugs:
*     {note_any_bugs_here}

*-

*  Type Definitions:
      IMPLICIT NONE              ! No implicit typing

*  Global Constants:
      INCLUDE 'SAE_PAR'          ! Standard SAE constants

*  Arguments Given:
      INTEGER EL

*  Arguments Returned:
      REAL ARRAY( * )

*  Status:
      INTEGER STATUS             ! Global status

*  Local Variables:
      INTEGER I                  ! Loop counter

*.

*  Check inherited global status.
      IF ( STATUS .NE. SAI__OK ) RETURN

*  Initialise the array.
      DO 1 I = 1, EL
         ARRAY( I ) = REAL( I )
 1    CONTINUE

      END

      SUBROUTINE SUM( EL, ARRAY, ISUM, STATUS )
*+
*  Name:
*     SUM

*  Purpose:
*     Sum the elements of an array.

*  Language:
*     Starlink Fortran 77

*  Invocation:
*     CALL SUM( EL, ARRAY, ISUM, STATUS )

*  Description:
*     Return the sum of the elements of a 1-dimensional array.

*  Arguments:
*     EL = INTEGER (Given)
*        Number of array elements.
*     ARRAY( EL ) = INTEGER (Given)
*        Array whose elements are to be summed.
*     ISUM = INTEGER (Returned)
*        Sum of array elements.
*     STATUS = INTEGER (Given and Returned)
*        The global status.

*  Authors:
*     RFWS: R.F. Warren-Smith (STARLINK, RAL)
*     {enter_new_authors_here}

*  History:
*     19-AUG-1991 (RFWS):
*        Original version.
*     {enter_changes_here}

*  Bugs:
*     {note_any_bugs_here}

*-

*  Type Definitions:
      IMPLICIT NONE              ! No implicit typing

*  Global Constants:
      INCLUDE 'SAE_PAR'          ! Standard SAE constants

*  Arguments Given:
      INTEGER EL
      INTEGER ARRAY( * )

*  Arguments Returned:
      INTEGER ISUM

*  Status:
      INTEGER STATUS             ! Global status

*  Local Variables:
      INTEGER I                  ! Loop counter

*.

*  Check inherited global status.
      IF ( STATUS .NE. SAI__OK ) RETURN

*  Initialise.
      ISUM = 0

*  Sum the array elements.
      DO 1 I = 1, EL
         ISUM = ISUM + ARRAY( I )
 1    CONTINUE

      END
